### PR TITLE
Bug #5341: HasMany via two relations

### PR DESCRIPTION
**Pull request by shirase**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #5341

### CODE REVIEW

File: framework/db/ActiveRelationTrait.php [hunk 1]
```diff
@@ -278,7 +279,14 @@ public function populateRelation($name, &$primaryModels)
             $buckets = $this->indexBuckets($buckets, $this->indexBy);
         }
 
-        $link = array_values(isset($viaQuery) ? $viaQuery->link : $this->link);
+        $link = array_values($this->link);
+        if (isset($viaQuery)) {
+            $deepViaQuery = $viaQuery;
+            while ($deepViaQuery->via) {
+                $deepViaQuery = is_array($deepViaQuery->via) ? $deepViaQuery->via[1] : $deepViaQuery->via;
```
Comment (dmirogin) [inline framework/db/ActiveRelationTrait.php]:
~~~php
is_array($deepViaQuery->via)
~~~
doesn't garantee that there is element with index 1

File: framework/db/ActiveRelationTrait.php [hunk 2]
```diff
@@ -380,21 +388,41 @@ private function populateInverseRelation(&$primaryModels, $models, $primaryName,
...
                 $key1 = $this->getModelKey($viaModel, $viaLinkKeys);
                 $key2 = $this->getModelKey($viaModel, $linkValues);
                 $map[$key2][$key1] = true;
             }
+
+            $viaQuery->viaMap = $map;
+
+            $viaVia = $viaQuery->via;
+            while ($viaVia) {
+                $viaViaQuery = is_array($viaVia) ? $viaVia[1] : $viaVia;
```
Comment (dmirogin) [inline framework/db/ActiveRelationTrait.php]: same

File: framework/db/ActiveRelationTrait.php [hunk 3]
```diff
@@ -380,21 +388,41 @@ private function populateInverseRelation(&$primaryModels, $models, $primaryName,
...
+                $viaMap = $viaViaQuery->viaMap;
+                $oldMap = $map;
+                $map = [];
+                foreach ($oldMap as $key1 => $keys) {
```
Comment (dmirogin) [inline framework/db/ActiveRelationTrait.php]: Don't like this way, key1, key2, key3, a lot of foreaches, can we do it better ?

**Review by dmirogin — DISMISSED** [review]

File: framework/db/ActiveRelationTrait.php (hunk 1 above)
Comment (shirase) [reply framework/db/ActiveRelationTrait.php]: Unnecessary checks. Format of `ActiveRelationTrait::via` property is fixed `[$relationName, $relation]` and must have index 1 if its type is Array. I can throw exception, but need add that check everywhere.

File: framework/db/ActiveRelationTrait.php (hunk 3 above)
Comment (shirase) [reply framework/db/ActiveRelationTrait.php]: I'll try to change

File: framework/db/ActiveRelationTrait.php [hunk 4]
```diff
@@ -423,6 +442,20 @@ private function buildBuckets($models, $link, $viaModels = null, $viaLink = null
         return $buckets;
     }
 
+    /**
+     * @param array $map
```
Comment (samdark) [inline framework/db/ActiveRelationTrait.php]: Need description

**Review by samdark — APPROVED** [review]
Looks good to me.

**Review by dmirogin — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by Faryshta** [comment]
Need another PR to update guides.

**Comment by samdark** [comment]
@Faryshta what should be updated? It is a bug that was fixed, not a new feature.

**Pull request merged**